# Marketing Analysis: Page Funnel Visits & A/B Testing

In this marketing analysis, we will delve into two crucial aspects: Page Funnel Visits and A/B Testing. By examining these topics, my aim to gain valuable insights into user behaviour, optimize the customer journey and enhance marketing strategies. To achieve this, I will utilize different datasets that provide comprehensive information on user interactions and experimental variations.

## Page Funnel Visits

The Page Funnel Visits project aims to analyze user behavior and conversion rates within a website's page funnel. A page funnel represents the sequential flow of user actions from initial page visits to specific conversion events, such as adding items to a cart, proceeding to checkout and making a purchase. By examining these user journeys, businesses can gain valuable insights into user engagement, identify potential bottlenecks and optimize the conversion process.

In [1]:
import pandas as pd

In [2]:
visits = pd.read_csv('visits.csv',
                     parse_dates=[1])
cart = pd.read_csv('cart.csv',
                   parse_dates=[1])
                   
checkout = pd.read_csv('checkout.csv',
                       parse_dates=[1])
purchase = pd.read_csv('purchase.csv',
                       parse_dates=[1])

In [3]:
print(visits.head(5))
print(cart.head(5))
print(checkout.head(5))
print(purchase.head(5))

                                user_id          visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00
                                user_id           cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00
1  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00
3  b594862a-36c5-47d5-b818-6e9512b939b3 2017-09-27 08:22:00
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974 2017-07-26 15:48:00
                                user_id       checkout_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00
1  4ac186f0-9954-4fea-8a27-c081e428e34e 2017-04-07 20:11:00
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d 2017-07-13 11:38:00
3  89fe330a-8966-4756-8f7c-3bdbcd47279a 

1. Page Visits:
   - Columns: user_id, visit_time
   - This dataset represents the visits made by users to a specific page on a website. It includes the unique identifier for each user (user_id) and the timestamp of their visit (visit_time).

2. Cart Time:
   - Columns: user_id, cart_time
   - This dataset records the time when users added items to their shopping carts. It includes the unique identifier for each user (user_id) and the timestamp when they added items to their cart (cart_time).

3. Checkout Time:
   - Columns: user_id, checkout_time
   - This dataset captures the checkout time of users who proceeded to the checkout process. It includes the unique identifier for each user (user_id) and the timestamp when they completed the checkout process (checkout_time).

4. Purchase Time:
   - Columns: user_id, purchase_time
   - This dataset contains information about the time when users made a purchase. It includes the unique identifier for each user (user_id) and the timestamp when they made the purchase (purchase_time).

These datasets provide a chronological flow of user activities, starting from page visits, followed by cart additions, checkout process and finally, purchases.

**Step 1: what percentage only visited?**

In [4]:
visits_cart = pd.merge(visits, cart, how = 'left')
visits_cart.head(10)

user_id          visit_time   
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00  \
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00   
5  7bc62113-d51d-4e6b-85e0-1cf29e3de74a 2017-06-03 20:05:00   
6  6f22cd44-bc63-4449-a9af-2651859da2f9 2017-03-18 23:09:00   
7  49a6399b-c095-4e42-92eb-af90abe2bb6a 2017-01-10 12:09:00   
8  653c095d-fd74-40dd-ae3b-5a69a2613629 2017-06-27 19:21:00   
9  3d9fdac1-aae3-4a01-b448-934d12047468 2017-09-24 06:43:00   

            cart_time  
0                 NaT  
1 2017-01-26 14:44:00  
2 2017-08-20 08:31:00  
3                 NaT  
4                 NaT  
5                 NaT  
6                 NaT  
7                 NaT  
8                 NaT  
9                 NaT

**How long is `visits_cart`?**

In [5]:
len(visits_cart)

2000

**How many timestamps are null for `cart_time`?**

In [6]:
len(visits_cart[visits_cart.cart_time.isnull()])

1652

In [7]:
len(visits_cart[visits_cart.cart_time.notnull()])/len(visits_cart)

0.174

The presence of null values in the "cart_time" column suggests that a significant number of users did not proceed to add items to their shopping carts after visiting the website. This could indicate a lack of interest or engagement with the products or a suboptimal user experience that discourages users from taking the next step.

**Step 2: what percentage placed a t-shirt in their cart?**

In [8]:
cart_checkout = pd.merge(cart, checkout, how = 'left')
len(cart_checkout[cart_checkout.checkout_time.notnull()])/len(cart_checkout)

0.6494252873563219

The fact that a significant majority of users who added a t-shirt to their cart went on to complete the checkout process is a positive sign. It suggests that the cart-to-checkout journey for t-shirt purchases is relatively successful, with a high conversion rate.

**Step 3: % of users who got to checkout?**

In [9]:
all_data = visits.merge(cart, how='left')\
    .merge(checkout, how='left')\
    .merge(purchase, how='left')
all_data.head(10)

user_id          visit_time   
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00  \
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00   
5  7bc62113-d51d-4e6b-85e0-1cf29e3de74a 2017-06-03 20:05:00   
6  6f22cd44-bc63-4449-a9af-2651859da2f9 2017-03-18 23:09:00   
7  49a6399b-c095-4e42-92eb-af90abe2bb6a 2017-01-10 12:09:00   
8  653c095d-fd74-40dd-ae3b-5a69a2613629 2017-06-27 19:21:00   
9  3d9fdac1-aae3-4a01-b448-934d12047468 2017-09-24 06:43:00   

            cart_time       checkout_time       purchase_time  
0                 NaT                 NaT                 NaT  
1 2017-01-26 14:44:00 2017-01-26 14:54:00 2017-01-26 15:08:00  
2 2017-08-20 08:31:00                 NaT                 NaT  
3                 NaT                 NaT                 NaT  
4                 NaT                 NaT                 NaT  
5                 NaT                 NaT                 NaT  
6                 NaT                 NaT                 NaT  
7                 NaT                 NaT                 NaT  
8                 NaT                 NaT                 NaT  
9                 NaT                 NaT                 NaT

In [10]:
len(all_data[all_data.purchase_time.notnull()])/len(all_data[all_data.checkout_time.notnull()])

0.7544910179640718

The high conversion rate suggests that the checkout process is effective in motivating users to complete their purchases. Users who reach the checkout stage are inclined to finalize their transactions, indicating that the checkout experience is user-friendly, offers a seamless process and instills confidence in users to complete their purchases.

**Checking each part of the funnel**

In [11]:
num1 = len(visits_cart[visits_cart.cart_time.notnull()])/len(visits_cart)
num2 = len(cart_checkout[cart_checkout.checkout_time.notnull()])/len(cart_checkout)
num3 = len(all_data[all_data.purchase_time.notnull()])/len(all_data[all_data.checkout_time.notnull()])
print(str(round(num1 * 100, 2)) + '%' + ' A user adds a t-shirt to their cart')
print(str(round(num2 * 100, 2)) + '%' + ' A user clicks “checkout”')
print(str(round(num3 * 100, 2)) + '%' + ' A user actually purchases a t-shirt')

17.4% A user adds a t-shirt to their cart
64.94% A user clicks “checkout”
75.45% A user actually purchases a t-shirt


*The weakest part of the funnel is clearly getting a person who visited the site to add a tshirt to their cart. Once they've added a t-shirt to their cart it is fairly likely they end up purchasing it. A suggestion could be to make the add-to-cart button more prominent on the front page.*

**Step 4: average time to purchase**

In [12]:
all_data['difference'] = all_data.purchase_time - all_data.visit_time

In [13]:
all_data.difference

0                  NaT
1      0 days 00:44:00
2                  NaT
3                  NaT
4                  NaT
             ...      
2103               NaT
2104               NaT
2105               NaT
2106               NaT
2107               NaT
Name: difference, Length: 2108, dtype: timedelta64[ns]

In [14]:
all_data.difference.mean()

Timedelta('0 days 00:43:12.380952380')

 The calculated average time to purchase provides insight into the typical duration users take from the checkout stage to completing their purchases. On average, users take approximately 43 minutes and 12 seconds to make their final purchase decision after reaching the checkout stage.

## A/B Testing

The online store is conducting an A/B test to compare the performance of two different versions of an ad. The ads are being displayed in emails, as well as in banner ads on popular social media platforms such as Facebook, Twitter and Google. The goal is to evaluate the performance of the ads across different platforms and analyze the variations observed on each day of the week. By analyzing the data, we can gain insights into which ad version and platform combination generates the most favorable outcomes for the online store.

In [15]:
ad_clicks = pd.read_csv('ad_clicks.csv')
ad_clicks

user_id utm_source            day   
0     008b7c6c-7272-471e-b90e-930d548bd8d7     google   6 - Saturday  \
1     009abb94-5e14-4b6c-bb1c-4f4df7aa7557   facebook     7 - Sunday   
2     00f5d532-ed58-4570-b6d2-768df5f41aed    twitter    2 - Tuesday   
3     011adc64-0f44-4fd9-a0bb-f1506d2ad439     google    2 - Tuesday   
4     012137e6-7ae7-4649-af68-205b4702169c   facebook     7 - Sunday   
...                                    ...        ...            ...   
1649  fe8b5236-78f6-4192-9da6-a76bba67cfe6    twitter     7 - Sunday   
1650  fed3db6d-8c92-40e3-a4fb-1fb9d7337eb1   facebook     5 - Friday   
1651  ff3a22ff-521c-478c-87ca-7dc7b8f34372    twitter  3 - Wednesday   
1652  ff3af0d6-b092-4c4d-9f2e-2bdd8f7c0732     google     1 - Monday   
1653  ffdfe7ec-0c74-4623-8d90-d95d80f1ba34   facebook   6 - Saturday   

     ad_click_timestamp experimental_group  
0                  7:18                  A  
1                   NaN                  B  
2                   NaN                  A  
3                   NaN                  B  
4                   NaN                  B  
...                 ...                ...  
1649                NaN                  A  
1650                NaN                  B  
1651                NaN                  B  
1652              22:57                  A  
1653                NaN                  B  

[1654 rows x 5 columns]

The provided data consists of the following columns:

1. `user_id`: A unique identifier for each user.
2. `utm_source`: The source of the ad, such as Google, Facebook or Twitter.
3. `day`: The day of the week when the ad was displayed.
4. `ad_click_timestamp`: The time when the ad was clicked. (NaN indicates no click)
5. `experimental_group`: The group to which the user belongs (A or B), representing the different versions of the ad.

The dataset contains 1654 rows with various user interactions across different platforms and days of the week. The "ad_click_timestamp" column indicates the specific time when an ad click occurred, and NaN values indicate instances where no click was made.

In [16]:
by_utm_source = ad_clicks.groupby('utm_source').user_id.count().reset_index()
by_utm_source

utm_source  user_id
0      email      255
1   facebook      504
2     google      680
3    twitter      215

These counts represent the number of users associated with each advertising source.

In [17]:
ad_clicks['is_click'] = ~ad_clicks\
   .ad_click_timestamp.isnull()
ad_clicks

user_id utm_source            day   
0     008b7c6c-7272-471e-b90e-930d548bd8d7     google   6 - Saturday  \
1     009abb94-5e14-4b6c-bb1c-4f4df7aa7557   facebook     7 - Sunday   
2     00f5d532-ed58-4570-b6d2-768df5f41aed    twitter    2 - Tuesday   
3     011adc64-0f44-4fd9-a0bb-f1506d2ad439     google    2 - Tuesday   
4     012137e6-7ae7-4649-af68-205b4702169c   facebook     7 - Sunday   
...                                    ...        ...            ...   
1649  fe8b5236-78f6-4192-9da6-a76bba67cfe6    twitter     7 - Sunday   
1650  fed3db6d-8c92-40e3-a4fb-1fb9d7337eb1   facebook     5 - Friday   
1651  ff3a22ff-521c-478c-87ca-7dc7b8f34372    twitter  3 - Wednesday   
1652  ff3af0d6-b092-4c4d-9f2e-2bdd8f7c0732     google     1 - Monday   
1653  ffdfe7ec-0c74-4623-8d90-d95d80f1ba34   facebook   6 - Saturday   

     ad_click_timestamp experimental_group  is_click  
0                  7:18                  A      True  
1                   NaN                  B     False  
2                   NaN                  A     False  
3                   NaN                  B     False  
4                   NaN                  B     False  
...                 ...                ...       ...  
1649                NaN                  A     False  
1650                NaN                  B     False  
1651                NaN                  B     False  
1652              22:57                  A      True  
1653                NaN                  B     False  

[1654 rows x 6 columns]

Now we have an additional column, "is_click," in the dataset. This column indicates whether a user clicked on the ad or not.

In [18]:
clicks_by_source = ad_clicks.groupby(['utm_source', 'is_click']).user_id\
  .count()\
  .reset_index()


clicks_pivot = clicks_by_source.pivot(
  columns = 'is_click',
  index = 'utm_source',
  values = 'user_id'
).reset_index()
clicks_pivot

is_click utm_source  False  True
0             email    175    80
1          facebook    324   180
2            google    441   239
3           twitter    149    66

In terms of user engagement, the "Google" platform stands out with the highest number of users who clicked on the ad (239). "Facebook" also shows good engagement, with 180 users clicking on the ad. "Twitter" and "Email" have relatively lower engagement rates, with 66 and 80 users clicking, respectively.

While the click-through rates provide insight into user engagement, it is also important to consider the total number of users associated with each platform. In this dataset, "Google" has the highest number of users (680), followed by "Facebook" (504), "Twitter" (215), and "Email" (255). These numbers suggest that "Google" has the potential to reach a larger audience and potentially drive more conversions.

In [19]:
clicks_pivot['percent_clicked'] = clicks_pivot[True] / (clicks_pivot[True] + clicks_pivot[False])
clicks_pivot

is_click utm_source  False  True  percent_clicked
0             email    175    80         0.313725
1          facebook    324   180         0.357143
2            google    441   239         0.351471
3           twitter    149    66         0.306977

Here are the click percentages based on the provided data:

1. `email`: Clicked - 31.4% (80 clicks out of 255 users)
2. `facebook`: Clicked - 35.7% (180 clicks out of 504 users)
3. `google`: Clicked - 35.1% (239 clicks out of 680 users)
4. `twitter`: Clicked - 28.0% (66 clicks out of 215 users)

These percentages represent the proportion of users who clicked on the ad out of the total number of users associated with each utm_source. They provide a more specific analysis of the click rates for each platform and allow for a comparison of their relative effectiveness. 

Based on these percentages, we can see that Facebook has the highest click percentage among the platforms analyzed, followed closely by Google. Twitter and Email have relatively lower click percentages. It's important to note that these percentages are based on the provided data and may change as more data is collected or when conducting further analysis.

In [20]:
a_or_b = ad_clicks.groupby('experimental_group').user_id.count().reset_index()
a_or_b

experimental_group  user_id
0                  A      827
1                  B      827

The dataset consists of two experimental groups, labeled as Group A and Group B. Each group contains an equal number of users, with 827 users in each group.

A/B test was designed with balanced sample sizes, ensuring an equal representation of users in both experimental groups. This balance is essential for a fair comparison between the two ad versions and helps minimize the impact of sample size imbalance on the test results.

In [21]:
a_or_b_by = ad_clicks.groupby(['experimental_group', 'is_click']).user_id.count().reset_index()

a_or_b_by_pivot = a_or_b_by.pivot(
  columns = 'is_click',
  index = 'experimental_group',
  values = 'user_id'
).reset_index()
a_or_b_by_pivot

is_click experimental_group  False  True
0                         A    517   310
1                         B    572   255

Comparing the number of "True" values in the "is_click" column, we can observe that Group A has 310 users who clicked on the ad, while Group B has 255 users who clicked on the ad. Group A had a higher click-through rate, with more users clicking on the ad. 

In [22]:
a_clicks = ad_clicks[
   ad_clicks.experimental_group
   == 'A'] 
b_clicks = ad_clicks[
   ad_clicks.experimental_group
   == 'B']

In [23]:
click_by_day_a = a_clicks.groupby(['is_click', 'day']).user_id.count().reset_index()
click_by_day_a_pivot = click_by_day_a.pivot(
  columns = 'is_click',
  index = 'day',
  values = 'user_id'
).reset_index()
click_by_day_a_pivot

is_click            day  False  True
0            1 - Monday     70    43
1           2 - Tuesday     76    43
2         3 - Wednesday     86    38
3          4 - Thursday     69    47
4            5 - Friday     77    51
5          6 - Saturday     73    45
6            7 - Sunday     66    43

For Group A:
- The highest number of clicks occurred on Friday (51 clicks), making it a favorable day for running ads in this group.
- Wednesday had the lowest number of clicks (38), indicating it may not be the most optimal day for ad performance.


In [24]:
click_by_day_b = b_clicks.groupby(['is_click', 'day']).user_id.count().reset_index()
click_by_day_b_pivot = click_by_day_b.pivot(
  columns = 'is_click',
  index = 'day',
  values = 'user_id'
).reset_index()
click_by_day_b_pivot

is_click            day  False  True
0            1 - Monday     81    32
1           2 - Tuesday     74    45
2         3 - Wednesday     89    35
3          4 - Thursday     87    29
4            5 - Friday     90    38
5          6 - Saturday     76    42
6            7 - Sunday     75    34

For Group B:
- Tuesday had the highest number of clicks (45), making it the most favorable day for running ads in this group.
- Saturday (42 clicks) and Friday (38 clicks) also showed relatively high click numbers, suggesting they could be effective days for ads as well.
- Thursday had the lowest number of clicks (29), indicating it may not be the most optimal day for ad performance.

Based on the click distribution data, both Friday and Saturday showed good results in terms of click numbers for their respective groups. In Group A, Friday had the highest number of clicks, while in Group B, Saturday had a relatively higher number of clicks compared to other days.

Considering this, it is recommended to allocate advertising efforts on both Friday and Saturday. These days have demonstrated potential for generating clicks and can be advantageous for running ads in both Group A and Group B. By targeting these days, we have a higher likelihood of capturing user attention and achieving better ad performance.